### DISCUSSION TITLE
Instance admins can see private keys?

### DISCUSSION THREAD

**Discussion by markcellus** [mastodon/mastodon#18669] [General]
I looked through the discussions and couldn't seem to find anything related to this. But it looks like Mastodon is storing private keys (from private/public keypair) in the instance's local database that can be seen by the admins. 

Sorry if I missed any previous discussions, but why was this decision made? Opposed to something more secure like allowing users sign and encrypt their private keys with a strong passphrase (they'd write down or store in a physical safe) before storing it in the database?

NOTE: I'm not a crypto expert so I may have the details wrong here. But I'd like to understand, from a high level, whether exposure of private keys to server admins can be addressed from a privacy perspective.

Thanks for any information!

**Comment by nightpool** [comment]
The "private keys" in Mastodon are a technical detail used to secure server to server communication. They have nothing to do with true end to end encryption, which Mastodon currently doesn't support (but is in progress). There are many design considerations that go into supporting true end to end encryption, especially in Mastodon's distributed model. The current trust model of Mastodon requires that you have *some* intermediary who hosts your content and acts as your gateway to the wider world, and it requires that you trust that intermediary to manage your account. Replacing that is much more complicated than "allow users to sign and encrypt their keys with a strong passphrase".

**Reply by markcellus** [reply]
Thanks for the response. But I'm not saying it's not complicated. I'm simply asking what was the rationale for the current behavior (if there was any at the time) vs any other alternative. If there is a link you can post, that would be awesome. Or are you saying the reason end-to-end encryption wasn't considered was because it was complicated?

**Reply by nightpool** [reply]
end-to-end encryption wasn't considered in the original design of Mastodon's protocol just because it's completely at odds with how every other web application works. Fundamentally, when you view a webpage, your root of trust for that domain is the person who is serving the web application. End-to-end encryption is valuable for users, but it is a completely different security model than the web in general is currently able to provide.

**Comment by Gargron** [comment]
Let me be clear here: The "private keys" that you are referring to have nothing to do with encryption. They are used for signing server-to-server communication. It might as well have been one private key shared between all users on a server, but it's not for various reasons.

**Reply by markcellus** [reply]
You and @nightpool keep saying that. I've never said Mastodon is using its private keys for encryption. I was simply asking about whether overall user privacy was taken into account when the app was made, and if there were any discussions about how to keep user's keys private from server admins. Encryption is just one way of possibly doing that but that was just a suggestion. Nightpool made it clear that there were never any prior discussions about privacy in this regard, which is fine.

But putting encryption aside, these "private keys" are still being used to authenticate requests to ensure messages via ActivityPub protocol are coming from and being sent to the right person, right? If that's the case, isn't it worth making these keys inaccessible to server admins? If not, couldn't a nefarious server admin make unauthenticated requests if they've gotten a hold of the private key?

**Reply by nightpool** [reply]
> about how to keep user's keys private from server admins.

No, what we're telling you is that because actor profile keys authenticate *server* communication, it is impossible to keep those keys ""private"" from the *server admin*. The entire concept makes zero sense technically. You would need a completely different key setup altogether.

> these "private keys" are still being used to authenticate requests to ensure messages via ActivityPub protocol are coming from and being sent to the right person, right

No. Again, like we've been saying, the keys are used to ensure that messages via the ActivityPub protocol are coming from the right *server*. They have nothing to do with authenticating an individual user. 

> couldn't a nefarious server admin make unauthenticated requests if they've gotten a hold of the private key?

A nefarious server admin could just change the posts that are hosted on their server *without bothering to do anything with the private keys at all*. The private keys are a complete distraction here, once you introduce a nefarious admin into the threat model, there are too many other possibilities to name.

**Reply by markcellus** [reply]
Yes, however new and less savvy Mastodon users do not understand the implications of signing up. I'm quite sure a non-tech user is not going to know whether an admin of the server they're using is nefarious or not. I'm also betting that the user also wouldn't assume that their server admin can take over their whole identity and send and accept messages to other instances on their behalf. This isn't an issue only for Mastodon, but I was just curious of whether you all thought about how to handle it.

>  the keys are used to ensure that messages via the ActivityPub protocol are coming from the right server. 

Ok, but if a bad actor owns multiple servers, they can essentially act on behalf of users, which to me seems like an issue that is worth addressing.

Look, I'm not trying to convince you all to prioritize user data and privacy if that's not something you want to do. Based on how defensive you're getting on this topic, it sounds like that's not a priority. And that's fine. I now understand your position on it. Thanks for the information.

**Reply by Gargron** [reply]
You are now clearly arguing in bad faith, and either wilfully or not still misunderstanding the presence of cryptographic keys in Mastodon. Their presence is completely irrelevant to the end-user experience. Think of them like SPF and DKIM keys in e-mail. The e-mail server has them, because they are about authenticating e-mail servers to each other, not the people behind e-mail addresses.
